sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/dynamodb/DynamoDBIO.java

```diff
@@ -0,0 +1,533 @@
+    }
+
+    public Read<T> withAwsClientsProvider(
+        String awsAccessKey, String awsSecretKey, String region, String serviceEndpoint) {
+      return withAwsClientsProvider(
+          new BasicDynamoDbProvider(awsAccessKey, awsSecretKey, region, serviceEndpoint));
+    }
+
+    public Read<T> withAwsClientsProvider(String awsAccessKey, String awsSecretKey, String region) {
+      return withAwsClientsProvider(awsAccessKey, awsSecretKey, region, null);
+    }
+
+    /**
+     * Can't pass ScanRequest object directly from client since this object is not full
+     * serializable.
+     */
+    public Read<T> withScanRequestFn(SerializableFunction<Void, ScanRequest> fn) {
+      return toBuilder().setScanRequestFn(fn).build();
+    }
+
+    private Read<T> withSegmentId(Integer segmentId) {
+      checkArgument(segmentId != null, "segmentId can not be null");
+      return toBuilder().setSegmentId(segmentId).build();
+    }
+
+    public Read<T> withScanResponseMapperFn(
+        SerializableFunction<ScanResponse, T> scanResultMapperFn) {
+      checkArgument(scanResultMapperFn != null, "scanResultMapper can not be null");
+      return toBuilder().setScanResponseMapperFn(scanResultMapperFn).build();
+    }
+
+    public Read<List<Map<String, AttributeValue>>> items() {
+      return withScanResponseMapperFn(new ItemsMapper())
+          .withCoder(ListCoder.of(MapCoder.of(StringUtf8Coder.of(), AttributeValueCoder.of())));
+    }
+
+    public Read<T> withCoder(Coder<T> coder) {
+      checkArgument(coder != null, "coder can not be null");
+      return toBuilder().setCoder(coder).build();
+    }
+
+    @Override
+    public PCollection<T> expand(PBegin input) {
+      checkArgument((getScanRequestFn() != null), "withScanRequestFn() is required");
+      checkArgument((getAwsClientsProvider() != null), "withAwsClientsProvider() is required");
+      ScanRequest scanRequest = getScanRequestFn().apply(null);
+      checkArgument(
+          (scanRequest.totalSegments() != null && scanRequest.totalSegments() > 0),
+          "TotalSegments is required with withScanRequestFn() and greater zero");
+
+      PCollection<Read<T>> splits =
+          (PCollection<Read<T>>)
+              input.apply("Create", Create.of(this)).apply("Split", ParDo.of(new SplitFn()));
+      splits.setCoder(SerializableCoder.of(new TypeDescriptor<Read<T>>() {}));
+
+      PCollection<T> output =
+          (PCollection<T>)
+              splits
+                  .apply("Reshuffle", Reshuffle.viaRandomKey())
+                  .apply("Read", ParDo.of(new ReadFn()));
+      output.setCoder(getCoder());
+      return output;
+    }
+
+    /** A {@link DoFn} to split {@link Read} elements by segment id. */
+    private static class SplitFn<T> extends DoFn<Read<T>, Read<T>> {
+      @ProcessElement
+      public void processElement(@Element Read<T> spec, OutputReceiver<Read<T>> out) {
+        ScanRequest scanRequest = spec.getScanRequestFn().apply(null);
```

Comment:
Do you want to get this merged first and then move to the `ScanSpec` API in a subsequent PR or maybe consider doing it as part of this PR as you prefer.

Reply:
@iemejia , in V2, we don't ScanSpec (nor QuerySpec)